### ISSUE TITLE
OpenVPN 2.3.11 and GUI not working with Duo

### ISSUE THREAD

**Issue by gena01** [OpenVPN/openvpn-gui#55]
I was trying to get the latest openvpn client 2.3.11 and the included GUI to work with user+password + duo (two factor) and somehow failed.

I have managed to run openvpn.exe by hand with `--auth-retry interact` and that worked, so it's something specific to the GUI.

The OpenVPN-GUI shows v10 version. Not sure what else I am missing.

**Comment by selvanair** [comment]
On Sat, Jun 18, 2016 at 10:05 PM, Gennady Feldman <notifications@github.com>
wrote:

> I was trying to get the latest openvpn client 2.3.11 and the included GUI
> to work with user+password + duo (two factor) and somehow failed.
>
> I have managed to run openvpn.exe by hand with --auth-retry interact and
> that worked, so it's something specific to the GUI.
>
> The OpenVPN-GUI shows v10 version. Not sure what else I am missing.
>

The stable releases of openvpn 2.3.x are bundled with GUI v10 which does
not support static challenge/response. You need version 11 for that.

Try one of the recent "daily" snapshot builds available at
http://build.openvpn.net/downloads/snapshots/
File names starting with "openvpn-install-master-yyyymmdd......." are the
ones built with git-master branch of openvpn and the current version of
GUI. Pick one of the latest builds for 64 bit (file name ending with
x86_64) or 32 bit (ending with i686) as needed. The ones dated June 14 are
the latest right now.

Selva

**Comment by gena01** [comment]
Are there plans:
- update the GUI in "stable" release(s)?
- update copyright date(s) in various places in the GUI app from 2005 to 2016?
- update/keep a Changes file. (CHANGES.rst only lists changes up to version 7)

**Comment by gena01** [comment]
Just tried the latest nightly (shows v11). Still not working :(

Log
```
Sat Jun 18 23:58:22 2016 VERIFY OK: depth=1, CN=OpenVPN CA
Sat Jun 18 23:58:22 2016 VERIFY OK: nsCertType=SERVER
Sat Jun 18 23:58:22 2016 VERIFY OK: depth=0, CN=OpenVPN Server
Sat Jun 18 23:58:22 2016 Data Channel Encrypt: Cipher 'BF-CBC' initialized with 128 bit key
Sat Jun 18 23:58:22 2016 Data Channel Encrypt: Using 160 bit message hash 'SHA1' for HMAC authentication
Sat Jun 18 23:58:22 2016 Data Channel Decrypt: Cipher 'BF-CBC' initialized with 128 bit key
Sat Jun 18 23:58:22 2016 Data Channel Decrypt: Using 160 bit message hash 'SHA1' for HMAC authentication
Sat Jun 18 23:58:22 2016 Control Channel: TLSv1, cipher TLSv1/SSLv3 ECDHE-RSA-AES256-SHA, 2048 bit RSA
Sat Jun 18 23:58:22 2016 [OpenVPN Server] Peer Connection Initiated with [AF_INET]52.23.101.91:1194
Sat Jun 18 23:58:24 2016 MANAGEMENT: >STATE:1466308704,GET_CONFIG,,,
Sat Jun 18 23:58:25 2016 SENT CONTROL [OpenVPN Server]: 'PUSH_REQUEST' (status=1)
Sat Jun 18 23:58:25 2016 AUTH: Received control message: AUTH_FAILED,CRV1:R,E:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx:zzzzzzz=:Duo passcode or second factor:
Sat Jun 18 23:58:25 2016 SIGUSR1[soft,auth-failure] received, process restarting
Sat Jun 18 23:58:25 2016 MANAGEMENT: >STATE:1466308705,RECONNECTING,auth-failure,,
Sat Jun 18 23:58:25 2016 Restart pause, 2 second(s)
```
I replaced the 2 strings in the AUTH_FAILED line

**Comment by selvanair** [comment]
On Sun, Jun 19, 2016 at 12:06 AM, Gennady Feldman <notifications@github.com>
wrote:

> Just tried the latest nightly (shows v11). Still not working :(
>
> Log
>
> Sat Jun 18 23:58:22 2016 VERIFY OK: depth=1, CN=OpenVPN CA
> Sat Jun 18 23:58:22 2016 VERIFY OK: nsCertType=SERVER
> Sat Jun 18 23:58:22 2016 VERIFY OK: depth=0, CN=OpenVPN Server
> Sat Jun 18 23:58:22 2016 Data Channel Encrypt: Cipher 'BF-CBC' initialized with 128 bit key
> Sat Jun 18 23:58:22 2016 Data Channel Encrypt: Using 160 bit message hash 'SHA1' for HMAC authentication
> Sat Jun 18 23:58:22 2016 Data Channel Decrypt: Cipher 'BF-CBC' initialized with 128 bit key
> Sat Jun 18 23:58:22 2016 Data Channel Decrypt: Using 160 bit message hash 'SHA1' for HMAC authentication
> Sat Jun 18 23:58:22 2016 Control Channel: TLSv1, cipher TLSv1/SSLv3 ECDHE-RSA-AES256-SHA, 2048 bit RSA
> Sat Jun 18 23:58:22 2016 [OpenVPN Server] Peer Connection Initiated with [AF_INET]52.23.101.91:1194
> Sat Jun 18 23:58:24 2016 MANAGEMENT: >STATE:1466308704,GET_CONFIG,,,
> Sat Jun 18 23:58:25 2016 SENT CONTROL [OpenVPN Server]: 'PUSH_REQUEST' (status=1)
> Sat Jun 18 23:58:25 2016 AUTH: Received control message: AUTH_FAILED,CRV1:R,E:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx:zzzzzzz=:Duo passcode or second factor:
> Sat Jun 18 23:58:25 2016 SIGUSR1[soft,auth-failure] received, process restarting
> Sat Jun 18 23:58:25 2016 MANAGEMENT: >STATE:1466308705,RECONNECTING,auth-failure,,
> Sat Jun 18 23:58:25 2016 Restart pause, 2 second(s)
>
> I replaced the 2 strings in the AUTH_FAILED line
>
Dynamic challenge is not supported by the GUI. At least not yet -- I have a
patch pending some cleanup, but it has not been a priority since no one
ever asked for dynamic challenge support in the GUI.

If your duo response is like sms/phone/push you can set it up to use the
static-challenge instead of dynamic challenge. Essentially, with static
challenge the client sends the username, password and an additional code
(could be an otp or a keyword like push) in the first phase of
authentication -- the GUI will prompt for them. In case of dynamic
challenge the first phase sends the username/password as usual and openvpn
responds with the dynamic challenge to which the client has to respond with
the second "password" -- the GUI does not understand this yet. Instead it
will show authentication failed and repeat the prompt for primary
username/password.

Selva

**Comment by selvanair** [comment]
If any one wants to try my working branch on dynamic response use the executable available here
https://github.com/selvanair/openvpn-gui/releases/tag/v11-dynamic-cr
Or build from the sources use the branch 'dynamic-cr' at https://github.com/selvanair/openvpn-gui.git

This is only the GUI so it requires an existing openvpn installation. The git-master branch of openvpn is preferred so that the interactive service may be used and the GUI may be run with limited user privileges.

This has been tested only against a test server. If you find any bugs please let me know.

**Comment by gena01** [comment]
Just tried the test build and it connected. :)

**Comment by selvanair** [comment]
The snapshot build you installed includes the service but the installer might not have setup it up. Open a  command prompt as admin and do this:

cd C:\Program Files\OpenVPN\bin
.\openvpnserv.exe -install
sc start OpenVPNServiceInteractive

**Comment by mattock** [comment]
Yeah, that is indeed the case: one needs to launch the Interactive Service manually. This will be fixed in later installers.

**Comment by selvanair** [comment]
Do you know what plugin your server is using for Duo? I wonder why it needs to use dynamic CR and not static CR. Response for Duo is always something fixed like push/sms/phone, isn't it?

**Comment by gena01** [comment]
I am not sure I wasn't the one who set it up. Duo is like Google Authenticator + push. So there's a code and/or push/sms/phone..

**Comment by selvanair** [comment]
That's why static response should work well.

One reason to use dynamic challenge may be to use a common config file for both users with and without 2-factor enabled. Then the server can decide to challenge to only those registered to use Duo. Just guessing. Anyway, I think this is the first time someone asked here for dynamic CR support.

**Comment by selvanair** [comment]
Dynamic CR support was added by commit 1a5ce44a99f927ca61b9c6faced4f0d99a3413b6
Included in openvpn release 2.4_rc1.
